Source/Evolve/WarpXEvolve.cpp

```diff
@@ -441,34 +441,40 @@ void WarpX::OneStep (
         }
         // electromagnetic solver
         else {
-            // perform particle collisions
-            ExecutePythonCallback("beforecollisions");
-            mypc->doCollisions(a_step, a_cur_time, a_dt);
-            ExecutePythonCallback("aftercollisions");
-
             // without mesh refinement
             if (finest_level == 0) {
                 // standard PIC loop
                 if (!m_JRhom) {
-                    OneStep_nosub(a_cur_time);
+                    OneStep_nosub(a_cur_time, a_dt, a_step);
                 }
                 // JRhom PIC loop
                 else {
+                    // perform particle collisions
+                    ExecutePythonCallback("beforecollisions");
+                    mypc->doCollisions(a_step, a_cur_time, a_dt);
+                    ExecutePythonCallback("aftercollisions");
+
                     OneStep_JRhom(a_cur_time);
                 }
             }
             // with mesh refinement
             else {
                 // without subcycling
                 if (!m_do_subcycling) {
-                    OneStep_nosub(a_cur_time);
+                    OneStep_nosub(a_cur_time, a_dt, a_step);
                 }
                 // with subcycling
                 else {
                     WARPX_ALWAYS_ASSERT_WITH_MESSAGE(
                         finest_level == 1,
                         "Subcycling not implemented with more than 1 mesh refinement level"
                     );
+
+                    // perform particle collisions
+                    ExecutePythonCallback("beforecollisions");
+                    mypc->doCollisions(a_step, a_cur_time, a_dt);
+                    ExecutePythonCallback("aftercollisions");
+
                     OneStep_sub1(a_cur_time);
```

Comment:
Note that in this PR I won't attempt to make the new algorithm work with the MR+subcycling method.